commit 130ea3973ba9ec0ebd767d6444268bfc33999489
Author: Katie Holly <email removed>
Date:   Fri Jul 4 17:51:03 2025 +0000

    Allow shrinking hashtables in low memory situations (#2095)
    
    During memory eviction, we may end up with a primary kv hash table
    that's below the 13% `MIN_FILL_PERCENT_SOFT`, but inside `resize()` for
    the kv table we check with `ht->type->resizeAllowed` whether allocating
    a new table crosses the configured `maxmemory` limit without taking into
    account that shrinking, while temporarily increasing the memory usage,
    will ultimately reduce the memory usage in all situations.
    
    Blocking hash tables from shrinking in situations where we would only
    temporarily cross the maxmemory limit has at least three negative side
    effects:
    * The performance of valkey is negatively impacted if, for example,
    we're moving items from table A to table B since table A would
    eventually end up being a very sparse hash table that never shrinks.
    * In extreme situations, the performance of valkey slows down
    significantly when trying to find more keys to evict since we're trying
    to find keys in a sparse table that only becomes more sparse over time.
    * In more extreme situations, such as when reducing `maxmemory` by
    around 90% or writing large data to a single key, valkey can end up in a
    situation where, no matter how many keys are being evicted from the kv
    table, the table bucket struct overhead uses more memory than the data
    stores in the table, causing all keys to be evicted from the table and
    valkey becoming unusable before setting maxmemory to above the current
    memory usage:
    
    ```
    127.0.0.1:6379> config set maxmemory 2gb
    OK
    127.0.0.1:6379> debug populate 20000000
    OK
    (8.94s)
    127.0.0.1:6379> debug htstats 0
    [Dictionary HT]
    Hash table 0 stats (main hash table):
     table size: 29360128
     number of entries: 20000000
    [Expires HT]
    127.0.0.1:6379> config set maxmemory 200mb
    OK
    127.0.0.1:6379> dbsize
    (integer) 18725794
    [...]
    127.0.0.1:6379> dbsize
    (integer) 0
    127.0.0.1:6379> keys *
    (empty array)
    127.0.0.1:6379> set foo bar
    (error) OOM command not allowed when used memory > 'maxmemory'.
    127.0.0.1:6379> config get maxmemory
    1) "maxmemory"
    2) "209715200"
    127.0.0.1:6379> info memory
    # Memory
    used_memory:269890904
    used_memory_human:257.39M
    [...]
    127.0.0.1:6379> config set maxmemory 300mb
    OK
    127.0.0.1:6379> info memory
    # Memory
    used_memory:1455512
    used_memory_human:1.39M
    [...]
    127.0.0.1:6379> set foo bar
    OK
    ```
    
    This PR addresses these issues by allowing hash tables to allocate a new
    table for shrinking, even if allocating a new table causes the maxmemory
    limit to be crossed during resize.
    
    Additionally, the check for whether we should fast-forward the resize
    operation has been moved to before checking whether we are allowed to
    allocate a new resize table. Because we don't start a resize in this
    section of the function yet, a fast-forward usually drops the memory
    usage enough to allow another resize operation to start immediately
    after the fast-forward resize completes.
    
    As a side effect of this PR, we now also refuse to start a hashtable
    resize if the policy is set to `HASHTABLE_RESIZE_FORBID`.
    
    ---------
    
    Signed-off-by: Fusl <email removed>
    Signed-off-by: Viktor Söderqvist <email removed>
    Co-authored-by: Viktor Söderqvist <email removed>

diff --git a/src/hashtable.c b/src/hashtable.c
index e2a956aca..eb64fd97d 100644
--- a/src/hashtable.c
+++ b/src/hashtable.c
@@ -614,14 +614,6 @@ static int resize(hashtable *ht, size_t min_capacity, int *malloc_failed) {
         return 0;
     }
 
-    if (ht->type->resizeAllowed) {
-        double fill_factor = (double)min_capacity / ((double)numBuckets(old_exp) * ENTRIES_PER_BUCKET);
-        if (fill_factor * 100 < MAX_FILL_PERCENT_HARD && !ht->type->resizeAllowed(alloc_size, fill_factor)) {
-            /* Resize callback says no. */
-            return 0;
-        }
-    }
-
     /* We can't resize if rehashing is already ongoing. Fast-forward ongoing
      * rehashing before we continue. This can happen only in exceptional
      * scenarios, such as when many insertions are made while rehashing is
@@ -633,6 +625,19 @@ static int resize(hashtable *ht, size_t min_capacity, int *malloc_failed) {
         }
     }
 
+    if (resize_policy == HASHTABLE_RESIZE_FORBID && ht->tables[0]) {
+        /* Refuse to resize if resizing is forbidden and we already have a primary table. */
+        return 0;
+    }
+    if (exp > old_exp && ht->type->resizeAllowed) {
+        /* If we're growing the table, let's check if the resizeAllowed callback allows the resize. */
+        double fill_factor = (double)min_capacity / ((double)numBuckets(old_exp) * ENTRIES_PER_BUCKET);
+        if (fill_factor * 100 < MAX_FILL_PERCENT_HARD && !ht->type->resizeAllowed(alloc_size, fill_factor)) {
+            /* Resize callback says no. */
+            return 0;
+        }
+    }
+
     /* Allocate the new hash table. */
     bucket *new_table;
     if (malloc_failed) {
@@ -1242,14 +1247,13 @@ int hashtableTryExpand(hashtable *ht, size_t size) {
 }
 
 /* Expanding is done automatically on insertion, but less eagerly if resize
- * policy is set to AVOID or FORBID. After restoring resize policy to ALLOW, you
- * may want to call hashtableExpandIfNeeded. Returns 1 if expanding, 0 if not
- * expanding. */
+ * policy is set to AVOID and not at all if set to FORBID.
+ * Returns 1 if expanding, 0 if not expanding. */
 int hashtableExpandIfNeeded(hashtable *ht) {
     size_t min_capacity = ht->used[0] + ht->used[1] + 1;
     size_t num_buckets = numBuckets(ht->bucket_exp[hashtableIsRehashing(ht) ? 1 : 0]);
     size_t current_capacity = num_buckets * ENTRIES_PER_BUCKET;
-    unsigned max_fill_percent = resize_policy == HASHTABLE_RESIZE_AVOID ? MAX_FILL_PERCENT_HARD : MAX_FILL_PERCENT_SOFT;
+    unsigned max_fill_percent = resize_policy == HASHTABLE_RESIZE_ALLOW ? MAX_FILL_PERCENT_SOFT : MAX_FILL_PERCENT_HARD;
     if (min_capacity * 100 <= current_capacity * max_fill_percent) {
         return 0;
     }
@@ -1257,21 +1261,34 @@ int hashtableExpandIfNeeded(hashtable *ht) {
 }
 
 /* Shrinking is done automatically on deletion, but less eagerly if resize
- * policy is set to AVOID and not at all if set to FORBID. After restoring
- * resize policy to ALLOW, you may want to call hashtableShrinkIfNeeded. */
+ * policy is set to AVOID and not at all if set to FORBID.
+ * Returns 1 if shrinking, 0 if not shrinking. */
 int hashtableShrinkIfNeeded(hashtable *ht) {
     /* Don't shrink if rehashing is already in progress. */
-    if (hashtableIsRehashing(ht) || resize_policy == HASHTABLE_RESIZE_FORBID || ht->pause_auto_shrink) {
+    if (hashtableIsRehashing(ht) || ht->pause_auto_shrink) {
         return 0;
     }
     size_t current_capacity = numBuckets(ht->bucket_exp[0]) * ENTRIES_PER_BUCKET;
-    unsigned min_fill_percent = resize_policy == HASHTABLE_RESIZE_AVOID ? MIN_FILL_PERCENT_HARD : MIN_FILL_PERCENT_SOFT;
+    unsigned min_fill_percent = resize_policy == HASHTABLE_RESIZE_ALLOW ? MIN_FILL_PERCENT_SOFT : MIN_FILL_PERCENT_HARD;
     if (ht->used[0] * 100 > current_capacity * min_fill_percent) {
         return 0;
     }
     return resize(ht, ht->used[0], NULL);
 }
 
+/* Resizes the hashtable to an optimal size, based on the current number of
+ * entries. This is a convenience function that first tries to shrink the table
+ * if needed, and then expands it if needed. After restoring resize policy
+ * to ALLOW, you may want to call hashtableShrinkIfNeeded.
+ * Returns 1 if resizing was performed, 0 otherwise. */
+int hashtableRightsizeIfNeeded(hashtable *ht) {
+    int ret = hashtableShrinkIfNeeded(ht);
+    if (!ret) {
+        ret = hashtableExpandIfNeeded(ht);
+    }
+    return ret;
+}
+
 /* Defragment the main allocations of the hashtable by reallocating them. The
  * provided defragfn callback should either return NULL (if reallocation is not
  * necessary) or reallocate the memory like realloc() would do.
diff --git a/src/server.c b/src/server.c
index 45891af04..2d17e8bc2 100644
--- a/src/server.c
+++ b/src/server.c
@@ -522,9 +522,6 @@ uint64_t dictEncObjHash(const void *key) {
 int hashtableResizeAllowed(size_t moreMem, double usedRatio) {
     UNUSED(usedRatio);
 
-    /* For debug purposes, not allowed to be resized. */
-    if (!server.dict_resizing) return 0;
-
     /* Avoid resizing over max memory. */
     return !overMaxmemoryAfterAlloc(moreMem);
 }
@@ -811,7 +808,7 @@ hashtableType clientHashtableType = {
  * for dict.c to resize or rehash the tables accordingly to the fact we have an
  * active fork child running. */
 void updateDictResizePolicy(void) {
-    if (server.in_fork_child != CHILD_TYPE_NONE) {
+    if (server.in_fork_child != CHILD_TYPE_NONE || !server.dict_resizing) {
         dictSetResizeEnabled(DICT_RESIZE_FORBID);
         hashtableSetResizePolicy(HASHTABLE_RESIZE_FORBID);
     } else if (hasActiveChildProcess()) {
